### PR TITLE
[Improvement][Alert] Add a test send feature when creating an alert instance

### LINKED ISSUE #15162: [Improvement][Alert] Add a test send feature when creating an alert instance
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Add a test send feature when creating an alert instance like the data source module.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qingwli**
- close #15162

image
image

### CODE REVIEW

File: dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java [hunk 1]
```diff
@@ -60,13 +61,17 @@ public static void main(String[] args) {
...
+
+        if (this.getClass().getName()
+                .contains(((SpringApplication) readyEvent.getSource()).getMainApplicationClass().getName())) {
+            log.info("Alert server is staring ...");
+            alertRegistryClient.start();
+            alertBootstrapService.start();
+            listenerEventPostService.start();
+            alertRpcServer.start();
+            log.info("Alert server is started ...");
+        }
```
Comment (ruanwenjun) [inline dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java]: Why do this change?

File: dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java [hunk 2]
```diff
@@ -260,4 +261,39 @@ public void close() {
         log.info("Closed ListenerEventPostService...");
     }
 
+    public AlertChannel getPluginDefine(int pluginDefineId) {
+        Optional<AlertChannel> alertChannelOptional = alertPluginManager.getAlertChannel(pluginDefineId);
+        return alertChannelOptional.orElse(null);
+    }
```
Comment (ruanwenjun) [inline dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java]:
```suggestion
    public Optional<AlertChannel> getPluginDefine(int pluginDefineId) {
        return  alertPluginManager.getAlertChannel(pluginDefineId);
    }
```

File: dolphinscheduler-api/pom.xml [hunk 3]
```diff
@@ -61,6 +61,11 @@
             <artifactId>dolphinscheduler-meter</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.dolphinscheduler</groupId>
+            <artifactId>dolphinscheduler-alert-server</artifactId>
+        </dependency>
+
```
Comment (ruanwenjun) [inline dolphinscheduler-api/pom.xml]: It's not a good idea to import alert-server in api-server, you should exposed a rpc method `testSend`  in alert server, and api call this method to test.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]
Good idea, leave some comments.

File: dolphinscheduler-ui/src/views/security/alarm-instance-manage/use-detail.ts [hunk 4]
```diff
@@ -41,6 +46,27 @@ export function useDetail(getFormValues: Function) {
...
+        pluginDefineId: values.pluginDefineId,
+        pluginInstanceParams: formatParams(json, values)
+      })
+      window.$message.success(
+        res
+          ? res.msg
+          : `${t('security.alarm_instance.test_send')} ${t('home.success')}`
+      )
+      status.testing = false
+    } catch (err) {
```
Comment (songjianet) [inline dolphinscheduler-ui/src/views/security/alarm-instance-manage/use-detail.ts]: please delete err, because it is not used.

**Review by songjianet — CHANGES_REQUESTED** [review]

File: dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java (hunk 2 above)
Comment (qingwli) [reply dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/service/ListenerEventPostService.java]: Done

File: dolphinscheduler-api/pom.xml (hunk 3 above)
Comment (qingwli) [reply dolphinscheduler-api/pom.xml]: Done

File: dolphinscheduler-ui/src/views/security/alarm-instance-manage/use-detail.ts (hunk 4 above)
Comment (qingwli) [reply dolphinscheduler-ui/src/views/security/alarm-instance-manage/use-detail.ts]: Done

File: dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java (hunk 1 above)
Comment (qingwli) [reply dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java]: Pre-design is to import the alert server in API-server, which will run the alert server auto, I just need the alert plugin manager and not need another component, so add this code. But change to use rpc does not need this change.

File: dolphinscheduler-extract/dolphinscheduler-extract-alert/src/main/java/org/apache/dolphinscheduler/extract/alert/IAlertOperator.java [hunk 5]
```diff
@@ -28,4 +29,7 @@ public interface IAlertOperator {
     @RpcMethod
     AlertSendResponse sendAlert(AlertSendRequest alertSendRequest);
 
+    @RpcMethod
+    AlertSendResponse.AlertSendResponseResult sendTestAlert(AlertTestSendRequest alertSendRequest);
```
Comment (ruanwenjun) [inline dolphinscheduler-extract/dolphinscheduler-extract-alert/src/main/java/org/apache/dolphinscheduler/extract/alert/IAlertOperator.java]:
```suggestion
    AlertSendResponse sendTestAlert(AlertTestSendRequest alertSendRequest);
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java [hunk 6]
```diff
@@ -352,4 +361,40 @@ private boolean checkHasAssociatedAlertGroup(String id) {
         return first.isPresent();
     }
 
+    @Override
+    public Result<Object> testSend(int pluginDefineId, String pluginInstanceParams) {
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java]:
```suggestion
    public Result<Void> testSend(int pluginDefineId, String pluginInstanceParams) {
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java [hunk 7]
```diff
@@ -352,4 +361,40 @@ private boolean checkHasAssociatedAlertGroup(String id) {
...
+        Result<Object> result = new Result<>();
+        Optional<Host> alertServerAddressOptional = registryClient.getAlertServerAddress();
+        if (!alertServerAddressOptional.isPresent()) {
+            log.error("Cannot get alert server address, please check the alert server is running");
+            putMsg(result, Status.ALERT_CHANNEL_NOT_EXIST);
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java]: Alert server not exist

File: docs/img/new_ui/dev/alert/alert_instance01.png
Comment (ruanwenjun) [inline docs/img/new_ui/dev/alert/alert_instance01.png:1]: Why change these pictures?

Comment (qingwli) [reply docs/img/new_ui/dev/alert/alert_instance01.png:1]: UI add `test send` button, so change the pic

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java (hunk 7 above)
Comment (qingwli) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java]: Done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java (hunk 6 above)
Comment (qingwli) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AlertPluginInstanceServiceImpl.java]: Done

File: dolphinscheduler-extract/dolphinscheduler-extract-alert/src/main/java/org/apache/dolphinscheduler/extract/alert/IAlertOperator.java (hunk 5 above)
Comment (qingwli) [reply dolphinscheduler-extract/dolphinscheduler-extract-alert/src/main/java/org/apache/dolphinscheduler/extract/alert/IAlertOperator.java]: Done

File: docs/img/new_ui/dev/alert/alert_instance01.png
Comment (qingwli) [reply docs/img/new_ui/dev/alert/alert_instance01.png:1]: Just pic03 contains this button, but pic03 shows table contains in pic1&2, so change those three pics

Comment (ruanwenjun) [reply docs/img/new_ui/dev/alert/alert_instance01.png:1]: I am not sure if the picture is too small.

Comment (qingwli) [reply docs/img/new_ui/dev/alert/alert_instance01.png:1]: Will be well 
image
image

**Review by ruanwenjun — DISMISSED** [review]
Backend LGTM

**Comment by qingwli** [comment]
@songjianet PTAL

File: docs/docs/en/guide/alert/alert_plugin_user_guide.md [hunk 8]
```diff
@@ -13,6 +13,8 @@ Steps to be used are as follows:
 - Select the corresponding alarm plug-in and fill in the relevant alarm parameters.
 - Select `Alarm Group Management`, create an alarm group, and choose the corresponding alarm instance.
 
+> You can click `Test send` button to test whether the alarm instance is configured correctly.
```
Comment (Radeity) [inline docs/docs/en/guide/alert/alert_plugin_user_guide.md]:
```suggestion
> You can click `Test Send` button to test whether the alarm instance is configured correctly.
```

**Review by Radeity — COMMENTED** [review]
Others LGTM

**Review by songjianet — APPROVED** [review]

**Pull request merged**